README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# edgarWebR

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/edgarWebR)](https://cran.r-project.org/package=edgarWebR)
![R-CMD-check](https://github.com/mwaldstein/edgarWebR/workflows/R-CMD-check/badge.svg)
[![codecov.io](https://codecov.io/github/mwaldstein/edgarWebR/coverage.svg?branch=master)](https://codecov.io/github/mwaldstein/edgarWebR?branch=master)

  - Author/Maintainer: [Micah J Waldstein](https://micah.waldste.in)
  - License: [MIT](https://opensource.org/licenses/MIT)

## Introduction

edgarWebR provides an interface to access the [SEC’s EDGAR
system](https://www.sec.gov/edgar/search-and-access) for company
financial filings.

edgarWebR does *not* provide any functionality to extract financial data
or other information from filings, only the metadata and company
information. For processing of the financial data.

## Ethical Use & Fair Access

Because of abusive use of this library, the SEC is likely to block its
use “as is” without setting a custom ‘User Agent’ identifier. Details
for setting a custom agent are below.

Users of this library are required to follow the SEC’s [Privacy and
Security Policy](https://www.sec.gov/privacy.htm#security). Failure to
follow that guidance may result in having your requests blocked. Per the
SEC’s [Developer Resources](https://www.sec.gov/developer):

> To ensure that everyone has equitable access to SEC EDGAR content,
> please use efficient scripting, downloading only what you need and
> please moderate requests to minimize server load. Current guidelines
> limit each user to a total of no more than 10 requests per second,
> regardless of the number of machines used to submit requests.
> 
> To ensure that SEC.gov remains available to all users, we reserve the
> right to block IP addresses that submit excessive requests. The SEC
> does not allow “unclassified” bots or automated tools to crawl the
> site. Any request that has been identified as part of an unclassified
> bot or an automated tool outside of the acceptable policy will be
> managed to ensure fair access for all users.

Users of this library are advised to set a custom user-agent by setting
the environment variable `EDGARWEBR_USER_AGENT`.

## EDGAR Tools

The EDGAR System provides a number of
[tools](https://www.sec.gov/edgar/search-and-access) for filing and
entity lookup and examination. As of v1.0, edgarWebR supports all public
search and browse interfaces.

*Search Interfaces:*

| Tool                          | URL                                                           | edgarWebR function(s)                                                                                                    |
| ----------------------------- | ------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| Company                       | <https://www.sec.gov/edgar/searchedgar/companysearch.html>    | `company_search()`, `company_information()`, `company_details()`, `company_filings()`                                    |
| Recent Filings                | <https://www.sec.gov/cgi-bin/browse-edgar?action=getcurrent>  | `latest_filings()`                                                                                                       |
| Full Text                     | <https://www.sec.gov/edgar/search/>                           | `full_text()`                                                                                                            |
| Header Search                 | <https://www.sec.gov/cgi-bin/srch-edgar>                      | `header_search()`                                                                                                        |
| Fund Disclosures              | <https://www.sec.gov/edgar/searchedgar/prospectus.htm>        | Use `company_search()` and specify the ‘type’ parameter as 485                                                           |
| Fund Voting Records           | <https://www.sec.gov/edgar/searchedgar/n-px.htm>              | Use `company_search()` and specify the ‘type’ parameter as ‘N-PX’                                                        |
| Fund Search                   | <https://www.sec.gov/edgar/searchedgar/mutualsearch.html>     | `fund_search()`, `fund_fast_search()`                                                                                    |
| Var. Insurance Products       | <https://www.sec.gov/edgar/searchedgar/vinsurancesearch.html> | `variable_insurance_search()`, `variable_insurance_fast_search()`                                                        |
| Confidential treatment orders | <https://www.sec.gov/edgar/searchedgar/ctorders.htm>          | Use `header_search()`, `company_search()`, `latest_filings()`, or `full_text()` and use form types ‘CT ORDER’            |
| Effectiveness notices         | <https://www.sec.gov/cgi-bin/browse-edgar?action=geteffect>   | `effectiveness()`                                                                                                        |
| CIK                           | <https://www.sec.gov/edgar/searchedgar/cik.htm>               | `cik_search()`                                                                                                           |
| Daily Filings                 | <https://www.sec.gov/edgar/searchedgar/currentevents.htm>     | `current_events()`                                                                                                       |
| Correspondence                | <https://www.sec.gov/answers/edgarletters.htm>                | Use `header_search()`, `company_search()`, `latest_filings()`, or `full_text()` and use form types ‘upload’ or ‘corresp’ |

Once a filing is found via any of the above, there are a number of
functions to process the result -

  - `filing_documents()`
  - `filing_filers()`
  - `filing_funds()`
  - `filing_information()`
  - `filing_details()` - returns all 4 of the filing components in a
    list.

### Parsing Tools

While edgarWebR is primarily focused on providing an interface to the
online SEC tools, there are a few activities for handling filing
documents for which no current tools exist.

  - `parse_submission()` - takes a full submission SGML document and
    parses out component documents. Most of the time, the documents of
    interest in a particular submission will be online and accessible
    via `filing_documents()` - this function is to unpack the raw
    submission to get all the documents. You may also find it more
    efficient if you’re regularly downloading all of the files in a
    given submission.
  - `parse_filing()` - Takes a HTML narrative filing and annotates each
    paragraph with item and part numbers.

### Data Sets

There is one dataset provided with edgarWebR - `sic_codes`, providing a
catalog of SIC codes and their hierarchy.

### URL Tools

There are also a number of utility functions to help construct useful
URL’s once you have a company CIK, submission accession number or
specific file.

  - `company_href()` for linking to the company page
  - `submission_index_href()` and its family of related functions for
    linking to a specific submission and file.

## Installation

edgarWebR is available from CRAN, so can be simply installed via

``` r
install.packages("edgarWebR")
```

To install the development version,

``` r
# Install the development version from GitHub:
# install.packages("devtools")
devtools::install_github("mwaldstein/edgarWebR")
```

## Example

``` r
company_filings("AAPL", type = "10-K", count = 10)
#>        accession_number act file_number filing_date accepted_date
#> 1  0000320193-19-000119  34   001-36743  2019-10-31    2019-10-30
#> 2  0000320193-18-000145  34   001-36743  2018-11-05    2018-11-05
#> 3  0000320193-17-000070  34   001-36743  2017-11-03    2017-11-03
#> 4  0001628280-16-020309  34   001-36743  2016-10-26    2016-10-26
#> 5  0001193125-15-356351  34   001-36743  2015-10-28    2015-10-28
#> 6  0001193125-14-383437  34   000-10030  2014-10-27    2014-10-27
#> 7  0001193125-13-416534  34   000-10030  2013-10-30    2013-10-29
#> 8  0001193125-12-444068  34   000-10030  2012-10-31    2012-10-31
#> 9  0001193125-11-282113  34   000-10030  2011-10-26    2011-10-26
#> 10 0001193125-10-238044  34   000-10030  2010-10-27    2010-10-27
#>                                                                                                href
#> 1  https://www.sec.gov/Archives/edgar/data/320193/000032019319000119/0000320193-19-000119-index.htm
#> 2  https://www.sec.gov/Archives/edgar/data/320193/000032019318000145/0000320193-18-000145-index.htm
#> 3  https://www.sec.gov/Archives/edgar/data/320193/000032019317000070/0000320193-17-000070-index.htm
#> 4  https://www.sec.gov/Archives/edgar/data/320193/000162828016020309/0001628280-16-020309-index.htm
#> 5  https://www.sec.gov/Archives/edgar/data/320193/000119312515356351/0001193125-15-356351-index.htm
#> 6  https://www.sec.gov/Archives/edgar/data/320193/000119312514383437/0001193125-14-383437-index.htm
#> 7  https://www.sec.gov/Archives/edgar/data/320193/000119312513416534/0001193125-13-416534-index.htm
#> 8  https://www.sec.gov/Archives/edgar/data/320193/000119312512444068/0001193125-12-444068-index.htm
#> 9  https://www.sec.gov/Archives/edgar/data/320193/000119312511282113/0001193125-11-282113-index.htm
#> 10 https://www.sec.gov/Archives/edgar/data/320193/000119312510238044/0001193125-10-238044-index.htm
#>    type film_number                                              form_name
#> 1  10-K   191181423 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 2  10-K   181158788 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 3  10-K   171174673 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 4  10-K   161953070 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 5  10-K   151180619 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 6  10-K   141175110 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 7  10-K   131177575 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 8  10-K   121171452 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 9  10-K   111159350 Annual report [Section 13 and 15(d), not S-K Item 405]
#> 10 10-K   101145250 Annual report [Section 13 and 15(d), not S-K Item 405]
#>    description  size
#> 1         <NA> 12 MB
#> 2         <NA> 12 MB
#> 3         <NA> 14 MB
#> 4         <NA> 13 MB
#> 5         <NA>  9 MB
#> 6         <NA> 12 MB
#> 7         <NA> 11 MB
#> 8         <NA>  9 MB
#> 9         <NA>  9 MB
#> 10        <NA> 13 MB
```

## Related Packages

  - [XBRL](https://CRAN.R-project.org/package=XBRL) - Low level
    extration of data from XBRL financial files
  - [finstr](https://github.com/bergant/finstr) - Process XBRL to
    extract data, combine periods, and make basic financial calulations.
  - [finreportr](https://github.com/sewardlee337/finreportr) - All in
    one to pull finnacials and information from EDGAR

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](https://mwaldstein.github.io/edgarWebR/CONDUCT.html). By
participating in this project you agree to abide by its terms. Report
violations to (<micah@waldste.in>).


NEWS.md
# edgarWebR 1.0.3.9999

## Features
 * Support new EDGAR full-text search
 * Add user-agent setting

## Internal
 * Change to use `with_mock_api` from httptest

# edgarWebR 1.0.3

## Bugfixes
 * Tests using dev testdata are skipped on cran to reduce overall package size

## Internal
 * Add github actions for testing to expand and normalize testing

# edgarWebR 1.0.2

## Bugfixes
 * User `options = 'HUGE'` for _all_ xml/html parsing - should resolve some
     test failures on Solaris.
 * All vignettes now use cached requests (tests were all ready using caches)

## Internal
 * httptest redaction/resolving logic centralized in a more standard way

# edgarWebR 1.0.1

## Bugfixes
 * `parse_text_filing()` correctly parses filings that have spaces in
     separating lines. (Fixes #4)
 * `parse_text_filing()` correctly removes and splits on '<PAGE>' dividers,
     including when there is no page number. (fixes #6)
 * Initial fix for #11 where some documents with especially deep structures
     would error in the parsing phase.
 * Check for errors connecting to SEC servers and stop on failures.

## Documentation
 * Elaborate the parsing vignette to better explain the files processed by
   `parse_filing()` and `parse_submission()`
 * Document output of `parse_submission()`
 * Document how the SEC Company 'Fast Search' works and how edgarWebR
   implements it. (addresses #7)

## Internal
 * Add missing pkgdown dev dependency
 * Add lintr dev dependency
 * Update tests w/ current data

# edgarWebR 1.0.0 - "Everything in its place"

## Features
 * `sic_codes` a dataset for sic codes from [US Department of
   Labor](https://www.osha.gov/pls/imis/sic_manual.html)
 * `parse_submission()` has options for skipping collection of all content or
   just binary content
 * Added utility functions `submission_index_href()`, `submission_href()`, and
   `submission_file_href()` for creating links to filings and their components.
 * Added `cik_search()` to lookup CIK codes for companies.
 * Added `current_events()` to access recent filings by form type
 * Added `company_search()`
 * Added `variable_insurance_search()` and `variable_insurance_fast_search()`
 * Added `fund_fast_search()`
 * Added `effectiveness()`

## Bugfixes
 * `parse_filing()` properly processes recent AIG & Costco filings
 * `parse_filing()` properly processes ADM filings for fy08 & fy09
 * `parse_filing()` now fails gracefully given a unparsable filing
 * `parse_filing()` properly handles extraneous '<PAGE>' tags from ~08/09 filings
   such as
   [ADP](https://www.sec.gov/Archives/edgar/data/8670/000120677409001642/adp_10k.htm)
 * `parse_filing()` handles more edge-cases and has similar word counts to plain
     document.
 * `parse_submission()` handles large files successfully using temp files.
 * `parse_submission()` handles submissions with single files correctly.
 * `parse_submission()` documented properly.
 * `full_text()` escapes the search query properly.

## Internal
 * Testing now requires tokenizers for testing if parsing adds words (it often
   does...)
 * Moved to testdata for parse_* functions rather than http cache
 * Made wordcount comparison in test_parse_filing far more restrictive (good
     thing!)
 * Tried to make TravisCI work, but still not great...
 * Added series_search to cover the main work for fund and variable insurance
     searches.

# edgarWebR 0.3.1 - "Oh yeah, data changes"
## Bugfixes
 * `full_text` no longer tests example as it often is long-running
 * `header_search` no longer tests example as it often is long-running
 * Vignette 'parsing' now always fetches the same filing to avoid issues with
   newer filings changing the output

# edgarWebR 0.3.0 - "Going all the way back"

## Features
 * `full_text` provides access to the full-text filing search interface.
 * `latest_filings` provides access to the latest SEC filings.
 * `parse_text_filing` parses text-only 10-* filings.
 * `parse_filing` detects when a filing is HTML wrapped plain text and uses
   `parse_text_filing` when appropriate.
 * `header_search` provides access to search filing headers back to 1994

## Bugfixes
 * `parse_filing` now treats `<br>` as a space avoiding words separated only
   by a line return getting concatenated. (Fixes #2)
 * `parse_filing` now tries to detect and remove TOC itemss/parts to avoid
   duplicate entries

## Internal
 * `map_xml` processes href's out of javascript links
 * `map_xml` add parameter for date format
 * linting setup to run in CI

# edgarWebR 0.2.1 - "Thanks Testers"

Many thanks to everyone providing feedback, particularly Günter Leitold for
rigorously testing the `parse_filing` funcion.

## Bugfixes
 * Parser covers a wider range of document formats for 10-K's.
 * `filing_documents` correctly provides the href for the document not the
   iXBRL viewer when present.

## Internal
 * Limited testing of `parse_filing` on CRAN.

# edgarWebR 0.2.0 - "Parse all the things"

## Features

### Parse Filings
New functionality in `parse_filing` to annotate a filing with part and item
identifiers. This enables easy extraction of risks or management discussions.

*NOTE:* While tested on a range of documents, due to the high variation in
filing formats, it is likely not to work on every document. Please open an
issue if errors are encountered.

### Parse Submission
New function `parse_submission` extracts documents from the full submission file.
This is needed for older submissions and accessing documents that are not extracted
individually online.

## Internal
 * Incorporated [httptest](https://github.com/nealrichardson/httptest) for cached testing. Altered most functions to use [httr](https://cran.r-project.org/package=httr) for remote calls to support this test method.
 * New Makefile targets to support cached testing -
   - `make test` - uses cached responses
   - `make test-nocache` - makes all remote calls
   - `make test-buildcache` - runs tests, caching remote requests.

R/utils.R
#' Submission URL Tools
#'
#' EDGAR submissions are organized fairly regularly. These functions help to
#' fint the URL to submission components.
#' @param cik Company code
#' @param accession accession number for a filing
#' @param filename filename provided in a submission
#' @return A string with URL requested
#' @examples
#' submission_index_href("0000712515", "0000712515-17-000090")
#' submission_href("0000712515", "0000712515-17-000090")
#' submission_file_href("0000712515", "0000712515-17-000090",
#'                      "pressrelease-ueberroth.htm")
#' @export
submission_index_href <- function(cik, accession) {
  submission_file_href(cik, accession, paste0(accession, "-index.htm"))
}

#' @describeIn submission_index_href Creates a link to the master submission
#' sgml submission file
#' @export
submission_href <- function(cik, accession) {
  submission_file_href(cik, accession, paste0(accession, ".txt"))
}

#' @describeIn submission_index_href provides the link to a given file within a
#' particular submission.
#' @export
submission_file_href <- function(cik, accession, filename) {
  trim_cik <- gsub("^0+", "", cik)
  dashless <- gsub("-", "", accession)
  paste0("https://www.sec.gov/Archives/edgar/data/", trim_cik, "/", dashless,
         "/", filename)
}

#' Company URL for a CIK
#'
#' Given a CIK, provide a link to the company information page.
#'
#' @param cik Company code
#' @param ownership (default: FALSE) boolean for inclusion of company change
#'   filings
#' @param atom (default: FALSE) if the link should be to the atom XML feed
#' @return A string with URL requested
#' @examples
#' company_href("0000037912")
#' @export
company_href <- function(cik, ownership = FALSE, atom = FALSE) {
  if (ownership == TRUE) ownership <- "include"
  if (ownership == FALSE) ownership <- "exclude"
  if (ownership == "") ownership <- "exclude"
  paste0(
    "https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany",
    "&CIK=", cik,
    "&owner=", ownership,
    "&hidefilings=0",
    ifelse(atom, "&output=atom", ""))
}

is_url <- function(x) {
  grepl("^(http|ftp)s?://", x, ignore.case = T)
}

get_doc <- function(x, clean = F) {
  if (typeof(x) == "character") {
    if (is_url(x)) {
      res <- edgar_GET(x)
      content <- httr::content(res, encoding = "UTF-8", as = "text")
      if (clean) {
        content <- clean_html(content)
      }
      # The 'HUGE' option can lead to some negative consequences with
      # particularly large documents, but given how poorly formed a lot of SEC
      # filings are, it is needed to ensure some parse at all...
      doc <- try({
        xml2::read_html(content, base_url = x)
      }, silent = T)
      if (inherits(doc, "try-error")) {
        doc <- xml2::read_html(content, base_url = x, options = "HUGE")
      }
    } else {
      if (clean) {
        content <- clean_html(x)
      } else {
        content <- x
      }
      doc <- try({
        xml2::read_html(content)
      }, silent = T)
      if (inherits(doc, "try-error")) {
        doc <- xml2::read_html(content, options = "HUGE")
      }
    }
  } else {
    doc <- x
  }

  if (clean) {
    doc <- clean_doc(doc)
  }

  doc
}

charToDoc <- function(x) {
  if (is_url(x)) {
    res <- edgar_GET(x)
    if (res$status != "200") {
      stop(paste0("Unable to reach the SEC endpoint (", x, ")"))
    }
    xml2::read_html(res, base_url = x, options = "HUGE")
  } else {
    xml2::read_html(x, options = "HUGE")
  }
}

charToText <- function(x) {
  if (is_url(x)) {
    res <- edgar_GET(x)
    if (res$status != "200") {
      stop(paste0("Unable to reach the SEC endpoint (", x, ")"))
    }
    return(httr::content(res, encoding = "UTF-8"))
  } else {
    return(x)
  }
}


unicode_map <- matrix(c(
  160,  " ",
  32,   " ",
  8194, " ", # En Space
  8195, " ", # Em Space
  8203, "",  # Zero-width space

  ## Hyphens
  151,  " - ",
  8208, " - ", # Hyphen
  8209, " - ", # Non-breaking Hyphen
  8210, " - ", # Figure Dash
  8211, " - ", # En dash
  8212, " - ", # Em dash
  8213, " - ", # Horizontal Bar
  8722, " - ", # Minus Sign

  ## Quotes
  145,  "'", # Private use One
  146,  "'", # Possessive Quote
  8216, "'", # Left Single Quote
  8217, "'", # Right Single Quote
  147,  "\"", # Set Transmit State (renered as double quote)
  148,  "\"", # Cancel Character (renered as double quote)
  8220, "\"", # Left Double Quote
  8221, "\"", # Right Double Quote

  ## Other
  8232, "\n", # Line Separator
  8260, "/",   # Fraction Slash
  "038", "&"
), ncol = 2, byrow = T)

html_escape_map <- unlist(apply(unicode_map, 1, function(map) {
  int.code <- as.integer(map[1])
  res <- list()
  res[paste0("&#", map[1], ";")] <- map[2]
  res[paste0("&#x", as.hexmode(int.code), ";")] <- map[2]
  res[paste0("&#x", toupper(as.hexmode(int.code)), ";")] <- map[2]
  res
}))

## This is a highly curated list based on what is actually seen in filings
## rather than an exhaustive code mapping
html_escape_map <- list(
  "&#8232;" = "\n",
  "&#x2011;" = " - ",
  "&#8210;"  = " - ",
  "&#x2212;" = " - ",
  "&#x2010;" = " - ",
  "&#150;"   = " - ",
  "&#151;"   = " - ",
  "&#8208;"  = " - ",
  "&#8209;"  = " - ",
  "&#8211;"  = " - ",
  "&#8212;"  = " - ",
  "&#8213;"  = " - ",
  "&#8722;"  = " - ",
  "&#x2013;" = " - ",
  "&#x2014;" = " - ",
  "&#x2015;" = " - ",
  "&#x2018;" = "'",
  "&#x2019;" = "'",
  "&#8216;"  = "'",
  "&#8217;"  = "'",
  "&#145;"   = "'",
  "&#146;"   = "'",
  "&#x201C;" = "\"",
  "&#x201D;" = "\"",
  "&#147;"   = "\"",
  "&#148;"   = "\"",
  "&#8220;"  = "\"",
  "&#8221;"  = "\"",
  "&#xa0;"    = " ",
  "&#32;"    = " ",
  "&#160;"   = " ",
  "&#x2003;" = " ",
  "&#8195;"  = " ",
  "&#8194;"  = " ",
  "&#8203;"  = " ", #zero width space
  "&#8260;"  = "/",
  "&#038;"   = "&"
)

html_escape_map["&nbsp;"] <- " "
html_escape_map["\u00a0"] <- " " # Unicode nbsp

# strips difficult to handle html bits we don't really care about
# @param x text of an html document
clean_html <- function(x) {
  # Not cleaned:
  # CODE, Count from SP500 Filings
  # "&#254;"         # thorn
  # "&#167;"         # sect
  # "&#174;"         # reg
  # "&#1465;",1      # Hebrew Point Holam
  # "&#8206;",8      # Left to Right mark
  # "&#8224;",8415   # Dagger
  # "&#8225;",1331   # Double Dagger
  # "&#8226;",178544 # Bullet
  # "&#8230;",655    # Ellipsis
  # "&#8232;",2      # Line Separator
  # "&#8356;",81     # Lira Sign
  # "&#8360;",2      # Rupee Sign
  # "&#8361;",10     # Won Sign
  # "&#8364;",3717   # Euro
  # "&#8369;",6      # Peso
  # "&#8480;",2      # Service Mark
  # "&#8482;",3914   # Trademark
  # "&#8539;",50     # Vulgar Fraction 1/8
  # "&#8540;",44     # Vulgar Fraction 3/8
  # "&#8541;",41     # Vulgar Fraction 5/8
  # "&#8542;",67     # Vulgar Fraction 7/8
  # "&#8718;",23     # End of Proof
  # "&#8729;",26     # Bullet Operator
  # "&#8730;",47     # Square Root
  # "&#8800;",2      # Not Equal To
  # "&#8804;",110    # Less-than or equal to
  # "&#8805;",273    # Greater-than or equal to
  # "&#8901;",5      # Dot operator
  # "&#9632;",37     # Black Square
  # "&#9633;",15     # White Square
  # "&#9642;",4846   # Black Square Small
  # "&#9675;",76     # White Circle
  # "&#9679;",4552   # Black Circle
  # "&#9702;",2029   # White Bullet
  # "&#9744;",1191   # Ballot Box
  # "&#9745;",397    # Ballot box w/ Check
  # "&#9746;",552    # Ballot Box w/ X
  # "&#9830;",130    # Black Diamond Suit
  # "&#9472;",36     # Circled Digit 0

  # character.replacements = list(
  #   ## SPACES
  #   "&nbsp;"  = " ",
  #   "&#160;"  = " ",
  #   "&#32;"   = " ",
  #   "\u00a0"  = " ", # Unicode nbsp
  #   "&#8194;" = " ", # En Space
  #   "&#8195;" = " ", # Em Space
  #   "&#8203;" = "",  # Zero-width space

  #   ## Hyphens
  #   "&#151;"  = " - ",
  #   "&#8208;" = " - ", # Hyphen
  #   "&#8209;" = " - ", # Non-breaking Hyphen
  #   "&#8210;" = " - ", # Figure Dash
  #   "&#8211;" = " - ", # En dash
  #   "&#8212;" = " - ", # Em dash
  #   "&#8213;" = " - ", # Horizontal Bar
  #   "&#8722;" = " - ", # Minus Sign

  #   ## Quotes
  #   "&#146;"  = "'", # Possessive Quote
  #   "&#8216;" = "'", # Left Single Quote
  #   "&#8217;" = "'", # Right Single Quote
  #   "&#8220;" = "\"", # Left Double Quote
  #   "&#8221;" = "\"", # Right Double Quote

  #   ## Other
  #   "&#8232;" = "\n", # Line Separator
  #   "&#8260;" = "/"   # Fraction Slash
  # )

  for (escape in names(html_escape_map)) {
    x <- gsub(escape,
              html_escape_map[escape],
              x,
              fixed = T)
      # x <- gsub(intToUtf8(substr(escape, 3, nchar(escape) - 1)),
      #           character.replacements[escape],
      #           x,
      #           fixed = T)
  }

  # xml_text doesn't break words on div closes, which we typically want
  x <- gsub("</div>", "</div> ", x, fixed = T)

  x <- gsub("<br>", " ", x, ignore.case = T)
  x <- gsub("<br/>", " ", x, ignore.case = T)
  x <- gsub("<page>", " ", x, ignore.case = T)
  x
}

# removes br from the doc since we don't really care about display, replacing
# with spaces
clean_doc <- function(doc) {
  replacement <- xml2::xml_find_first(xml2::read_xml("<p> </p>"),
                                      "/p/text()")
  xml2::xml_replace(xml2::xml_find_all(doc, "//br"), replacement)

  # remove hidden divs
  xml2::xml_remove(xml2::xml_find_all(doc, "//div[@style = 'display:none']"),
                   free = T)

  # Don't care about non-text divs
  xml2::xml_remove(xml2::xml_find_all(doc, "//div[(count(*) = 0 or count(hr) =
                                      count(*)) and normalize-space() = '']"), free = T)

  # strip messy inlineXBRL
  if (length(xml2::xml_ns(doc)) > 1) {
    xml2::xml_remove(xml2::xml_find_all(doc, "//header"), free = T)
  }

  doc
}
